### PR TITLE
Update to stable kotlin dsl

### PR DESCRIPTION
**Pull request by marcoferrer**
This PR updates the Kotlin DSL version to stable  ~~1.0.3~~ 1.0.4 (version included in gradle 5.0) and resolves https://github.com/google/protobuf-gradle-plugin/issues/274

Unfortunately it breaks compatibility with gradle 4.10.2 and ktdsl 1.0.0-rc6, due to changes in constructor visibility in ```NamedDomainObjectContainerScope```.  The release version exposes a static factory method for object instantiation that is not present in rc6. 

 Im still investigating the possibility of maintaining compatibility.

---
This change is [Reviewable](https://reviewable.io/reviews/google/protobuf-gradle-plugin/275)

### CODE REVIEW

File: testProjectKotlinDslBase/build.gradle.kts [hunk 1]
```diff
@@ -100,7 +100,7 @@ tasks {
             from(sourceSet.output)
 
             val compileTaskName = sourceSet.getCompileTaskName("java")
-            dependsOn(tasks.getByName(compileTaskName))
+            dependsOn(project.tasks.getByName(compileTaskName))
```
Comment (marcoferrer) [inline testProjectKotlinDslBase/build.gradle.kts]: An additional extension was added to the task container in 1.0.3. So in order to resolve ambiguity in method resolution I had to explicitly qualify the ```tasks``` property.

File: src/main/kotlin/com/google/protobuf/gradle/ProtobufConfiguratorExts.kt [hunk 2]
```diff
@@ -24,7 +24,7 @@ fun ProtobufConfigurator.protoc(closure: ExecutableLocator.() -> Unit) {
 
 fun ProtobufConfigurator.plugins(closure: NamedDomainObjectContainerScope<ExecutableLocator>.() -> Unit) {
     plugins(closureOf<NamedDomainObjectContainer<ExecutableLocator>> {
-        this(closure)
+        closure(NamedDomainObjectContainerScope.of(this))
```
[the remark this answers is not in the document]
Comment (alexvas) [reply src/main/kotlin/com/google/protobuf/gradle/ProtobufConfiguratorExts.kt]: In Kotlin `NamedDomainObjectContainerScope<ExecutableLocator>` on the 2nd line is so-called _receiver_.

In `closure: NamedDomainObjectContainerScope<ExecutableLocator>.() -> Unit` closure parameter is a kind of 'extension' method, that would be executed on the appropriate receiver.

Back from bottom to top:

1.
```kotlin
// extension method needs a receiver to be executed upon. The receiver is created
// by factory method NamedDomainObjectContainerScope.of(this)
closure(NamedDomainObjectContainerScope.of(this))
```
2.
```kotlin
// here is what _this_ in the bottom line is. It is again a 'receiver' (now in groovy sense)
// of type NamedDomainObjectContainer<ExecutableLocator>
(closureOf<NamedDomainObjectContainer<ExecutableLocator>> {
```
3. 
and when we look inside 'plugins' definition:
```groovy
  public void plugins(Closure configureClosure) {
    ConfigureUtil.configure(configureClosure, tools.plugins)
  }
```
indeed, tools.plugins parameter is of type NamedDomainObjectContainer<ExecutableLocator>.

So, back from top to bottom. The parameter tools.plugins of type NamedDomainObjectContainer<ExecutableLocator> would be configured with closureOf, it becomes _this_ in the bottom line, where it is used as a factory method parameter. Factory method instantiates an object that would be configured by _closure_ parameter of the top line.

**Comment by marcoferrer** [comment]
@zhangkun83 Sorry about the delay. Im updating the gradle version and looking into the failures now.

File: build.gradle [hunk 3]
```diff
@@ -83,6 +86,8 @@ dependencies {
     }
     testCompile 'commons-io:commons-io:2.5'
 
+    // This kotlin version needs to be compatible with the 
+    // android plugin being used in the test runtime.
     testProjectRuntime "org.jetbrains.kotlin:kotlin-gradle-plugin:1.2.0"
```
Comment (marcoferrer) [inline build.gradle]: Bumping this dependency version is what was causing the failures in the ci. The test runtime kotlin version needs to be compatible with the older version of the android plugin. I didnt need to update it in the first place so its fine being set to ```1.2.0```

File: src/main/kotlin/com/google/protobuf/gradle/ProtobufConfiguratorExts.kt [hunk 4]
```diff
@@ -4,61 +4,170 @@ import org.gradle.api.NamedDomainObjectContainer
...
+ */
+fun ProtobufConfigurator.protoc(action: ExecutableLocator.() -> Unit) {
+    protoc(closureOf(action))
 }
 
-fun ProtobufConfigurator.plugins(closure: NamedDomainObjectContainerScope<ExecutableLocator>.() -> Unit) {
+fun ProtobufConfigurator.plugins(action: NamedDomainObjectContainerScope<ExecutableLocator>.() -> Unit) {
     plugins(closureOf<NamedDomainObjectContainer<ExecutableLocator>> {
-        this(closure)
+        NamedDomainObjectContainerScope.of(this).action()
```
Comment (marcoferrer) [inline src/main/kotlin/com/google/protobuf/gradle/ProtobufConfiguratorExts.kt]: I changed the syntax usage in these methods to improve readability. At first I used the syntax from the code generated from gradle's plugin resolution stage for kotlin dsl. In hind sight, it was confusing to interpret for those not familiar with kotlin.

File: src/main/kotlin/com/google/protobuf/gradle/ProtobufConfiguratorExts.kt [hunk 5]
```diff
@@ -4,61 +4,170 @@ import org.gradle.api.NamedDomainObjectContainer
...
+ *
+ * @receiver [NamedDomainObjectContainerScope] The scope of the [NamedDomainObjectContainer]
+ * on which to create or configure an element.
+ *
+ * @param id The string id of the element to create or configure.
+ * @param action An optional action that will be applied to the element instance.
+ *
+ * @return [Unit]
+ */
+fun <T : Any> NamedDomainObjectContainerScope<T>.id(id: String, action: (T.() -> Unit)? = null) {
```
Comment (marcoferrer) [inline src/main/kotlin/com/google/protobuf/gradle/ProtobufConfiguratorExts.kt]: These two methods arent new. I just moved them down in the file and documented their usage

**Comment by marcoferrer** [comment]
@zhangkun83 CI issues have been resolved. I also updated a few things for readability and added some documentation for what I can at the moment.

File: README.md [hunk 6]
```diff
@@ -53,8 +53,8 @@ buildscript {
 Stand-alone examples are available for each of gradle's supported languages.
  * [Groovy](https://github.com/google/protobuf-gradle-plugin/tree/master/examples/exampleProject) ___(Default)___
    * Run `../../gradlew build` under the example directory to test it out.
- * [Kotlin (experimental)](https://github.com/google/protobuf-gradle-plugin/tree/master/examples/exampleKotlinDslProject)
-   * Run `./gradlew build` under the Kotlin example directory to test it out. This example is set up with Gradle 4.10, the minimum required version.
+ * [Kotlin](https://github.com/google/protobuf-gradle-plugin/tree/master/examples/exampleKotlinDslProject) ___(Experimental)___
+   * Run `./gradlew build` under the Kotlin example directory to test it out. This example is set up with Gradle 5.0, the minimum required version.
```
Comment (zhangkun83) [inline README.md]: Probably worth speaking louder about this requirement.

**Comment by marcoferrer** [comment]
So I think I found a work around for the compatibility problem between 4.10 and 5.0, but I did notice that master was failing with the same issue as my branch.
```
Caused by: org.gradle.internal.resolve.ModuleVersionNotFoundException: Could not find com.android.tools.build:gradle:2.2.0.
    Searched in the following locations:
        https://dl.google.com/dl/android/maven2/com/android/tools/build/gradle/2.2.0/gradle-2.2.0.pom
        https://dl.google.com/dl/android/maven2/com/android/tools/build/gradle/2.2.0/gradle-2.2.0.jar
        https://jcenter.bintray.com/com/android/tools/build/gradle/2.2.0/gradle-2.2.0.pom
        https://jcenter.bintray.com/com/android/tools/build/gradle/2.2.0/gradle-2.2.0.jar
        https://plugins.gradle.org/m2/com/android/tools/build/gradle/2.2.0/gradle-2.2.0.pom
        https://plugins.gradle.org/m2/com/android/tools/build/gradle/2.2.0/gradle-2.2.0.jar
```

File: src/main/groovy/com/google/protobuf/gradle/KtDslCompatibilityUtils.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+     * to which ever kotlin-dsl ext implementation is available on the classpath.
+     *
+     * Since NamedDomainObjectContainerExtensionsKt.invoke is an inline extension function, our
+     * usages in 'ProtobufConfiguratorExts.kt' would use the byte code for which ever kotlin-dsl
+     * version we compiled against. This caused issues with providing compatibility with
+     * kotlin-dsl 1.0.0-rc3 and 1.0.4 (Stable).
+     *
+     * Since the kotlin compiler creates a static implementations of all extensions, we can create a
+     * delegating utility function that ensures we are not inline-ing our kotlin-dsl compilation
+     * target byte code.
```
Comment (zhangkun83) [inline src/main/groovy/com/google/protobuf/gradle/KtDslCompatibilityUtils.java]: Sorry, I still don't understand why this additional layer of indirection would prevent the problem. Doesn't this method inline the `invoke()` call too?

**Comment by alexvas** [comment]
As of plugin version of 0.8.8 I still need a pair of auxiliary functions: 

```kotlin
fun ProtobufConfigurator.plugins(closure: NamedDomainObjectContainerScope<ExecutableLocator>.() -> Unit) {
    plugins(closureOf<NamedDomainObjectContainer<ExecutableLocator>> {
        closure(NamedDomainObjectContainerScope.of(this))
    })
}

fun GenerateProtoTask.plugins(configuration: NamedDomainObjectContainerScope<GenerateProtoTask.PluginOptions>.()-> Unit) {
    plugins(closureOf<NamedDomainObjectContainer<GenerateProtoTask.PluginOptions>> {
        configuration(NamedDomainObjectContainerScope.of(this))
    })
}
```
to make

```kotlin
protobuf {
    protoc {
        // The artifact spec for the Protobuf Compiler
        artifact = Libs.protoc
    }
    plugins {
        // Optional: an artifact spec for a protoc plugin, with "grpc" as
        // the identifier, which can be referred to in the "plugins"
        // container of the "generateProtoTasks" closure.
        id("grpc") {
            artifact = Libs.protoc_gen_grpc_java
        }
    }
   generateProtoTasks {
        ofSourceSet("main").forEach {
            it.plugins {
                // Apply the "grpc" plugin whose spec is defined above, without
                // options.  Note the braces cannot be omitted, otherwise the
                // plugin will not be added. This is because of the implicit way
                // NamedDomainObjectContainer binds the methods.
                id("grpc")
            }
        }
    }
}
```
work. Otherwise there is a gradle configuration error:

```
  Line 31:         id("grpc") {
                              ^ Too many arguments for public open fun id(id: String): PluginDependencySpec defined in org.gradle.kotlin.dsl.PluginDependenciesSpecScope

  Line 32:             artifact = Libs.protoc_gen_grpc_java
                       ^ Unresolved reference: artifact

2 errors
```

**Pull request merged**